### PR TITLE
Reader Card Refresh: Stack media vertically and make full width

### LINKED ISSUE #66621: Reader Card Refresh: Stack media vertically and make full width
## Description

We're in the process of updating the look and feel of our Calypso Reader Cards. As a part of this, we'd like to make all media full-width vs left aligned. We'll want to make updates to the following cards:

- Single image
- Two images
- Three images
- Four images
- Video
- Gallery

### Before
Image

#### Single Image
Image

#### Two image
Image

#### Three image
Image

#### Four image
Image

#### Video
Image

#### Gallery
Image

### PR DESCRIPTION
**Pull request by eoigal**
## Description

We're in the process of updating the look and feel of our Calypso Reader Cards. As a part of this, we'd like to make all media full-width vs left aligned. We'll want to make updates to the following cards:

- Single image
- Two images
- Three images
- Four images
- Video
- Gallery

### Before
Image

#### Single Image
Image

#### Two image
Image

#### Three image
Image

#### Four image
Image

#### Video
Image

#### Gallery
Image

### Changes of note in this PR

* Added a featured images component to show between 1-4 images from a post
* Created flex containers to allow featured images component to display images to match mockups
* Updated the gallery to show 5 - 10 images in full width slideshow format
* Fixed bug in video embeds, where videopress thumbnails were not rendering
* 
*

### CODE REVIEW

**Comment by davemart-in** [comment]
First review. Noticed a couple of things:

1) Recommended posts (2 posts next to each other) are currently escaping the grid. Both pictures should fit in the same width as the single image:

CleanShot 2022-09-06 at 10 58 54@2x

2) I did a search for "videopress" and noticed that posts with VideoPress video embeds show up with a black background (no thumbnail):

CleanShot 2022-09-06 at 11 03 19@2x

cc @ollierozdarz in case you have any additional feedback. Note: You should be able to click the "Calypso Live" link at the top of this PR to preview these changes:

CleanShot 2022-09-06 at 11 05 17@2x

**Comment by eoigal** [comment]
I think there is an issue with Videopress thumbnails (not having one). When I look at search results for youtube the thumbnail of the video is rendering ok. I'll debug further.
Screenshot 2022-09-06 at 18 32 02

**Comment by eoigal** [comment]
I've fixed the issue with the related posts width and I've reached out to Videopress for more details on why we receive a 403 when trying to access videopress thumbnail URLs - p1662543833203299-slack-C02LT75D3

File: client/blocks/reader-post-card/featured-images.jsx [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+			</ul>
+		);
+	}
+
+	return <ul className="reader-post-card__featured-images">{ listItems }</ul>;
+};
+
+ReaderFeaturedImages.propTypes = {
+	post: PropTypes.object.isRequired,
+	isDiscover: PropTypes.bool,
```
Comment (roo2) [inline client/blocks/reader-post-card/featured-images.jsx]: isDiscover doesn't seem to be used by ReaderFeaturedImages(?)

File: client/blocks/reader-featured-image/index.jsx [hunk 2]
```diff
@@ -2,56 +2,69 @@ import classnames from 'classnames';
...
 	children,
 	onClick,
 	className,
 	fetched,
+	imageUrl,
+	imageWidth,
+	imageHeight,
 } ) => {
-	if ( imageUrl === undefined ) {
+	const featuredImageUrl = imageUrl || canonicalMedia.src;
```
Comment (roo2) [inline client/blocks/reader-featured-image/index.jsx]: nit: I think I'd do a null safe check for `canonicalMedia.src` I see the prop is _required_ but I don't think that is really enforced.
```suggestion
	const featuredImageUrl = imageUrl || canonicalMedia?.src;
```

File: client/blocks/reader-featured-image/index.jsx [hunk 3]
```diff
@@ -2,56 +2,69 @@ import classnames from 'classnames';
...
+	featuredImageStyle.width = imageWidth || READER_CONTENT_WIDTH;
 
+	const divStyle = { height: newHeight };
 	const classNames = classnames( className, 'reader-featured-image' );
 
 	return (
-		<a className={ classNames } href={ href } style={ featuredImageStyle } onClick={ onClick }>
-			{ children }
-		</a>
+		<div style={ divStyle }>
```
Comment (roo2) [inline client/blocks/reader-featured-image/index.jsx]: nit: I don't understand why the div is now necessary?

File: client/blocks/reader-post-card/gallery.jsx [hunk 4]
```diff
@@ -1,66 +1,111 @@
...
+		event.preventDefault();
+		let updateItemSelected;
+		updateItemSelected = this.state.itemSelected + 1;
+		if ( updateItemSelected >= this.state.listItems.length ) {
+			updateItemSelected = 0;
 		}
+		this.setState( { itemSelected: updateItemSelected } );
+	};
+
+	getGalleryWorthyImages = ( post ) => {
```
Comment (roo2) [inline client/blocks/reader-post-card/gallery.jsx]: nit: it looks like this could be pulled out somewhere into a helper function

File: client/blocks/reader-post-card/gallery.jsx [hunk 5]
```diff
@@ -1,66 +1,111 @@
...
-			imageStyle = {
-				backgroundImage: 'url(' + safeCssUrl + ')',
-				backgroundSize: 'cover',
-				backgroundPosition: '50% 50%',
-				backgroundRepeat: 'no-repeat',
-			};
+	handleClick = ( event ) => {
+		event.preventDefault();
+		let updateItemSelected;
+		updateItemSelected = this.state.itemSelected + 1;
```
Comment (roo2) [inline client/blocks/reader-post-card/gallery.jsx]: nit: This could be one line

File: client/blocks/reader-post-card/gallery.jsx [hunk 6]
```diff
@@ -1,66 +1,111 @@
...
+			const safeCssUrl = cssSafeUrl( imageUrl );
+			let imageStyle = { background: 'none' };
+			if ( safeCssUrl ) {
+				imageStyle = {
+					backgroundImage: 'url(' + safeCssUrl + ')',
+					backgroundSize: 'cover',
+					backgroundPosition: '50% 50%',
+					backgroundRepeat: 'no-repeat',
+				};
+			}
```
Comment (roo2) [inline client/blocks/reader-post-card/gallery.jsx:42]: nit: this could potentially also be a helper function shared with reader-featured-images.jsx

File: client/blocks/reader-post-card/index.jsx [hunk 7]
```diff
@@ -139,7 +139,7 @@ class ReaderPostCard extends Component {
 		if ( isEligibleForUnseen( { isWPForTeamsItem, currentRoute, hasOrganization } ) ) {
 			isSeen = post?.is_seen;
 		}
-		const isPhotoPost = !! ( post.display_type & DisplayTypes.PHOTO_ONLY ) && ! compact;
+		const isPhotoPost = false; //!! ( post.display_type & DisplayTypes.PHOTO_ONLY ) && ! compact;
```
Comment (roo2) [inline client/blocks/reader-post-card/index.jsx]: commented out code

File: client/blocks/reader-post-card/gallery.jsx [hunk 8]
```diff
@@ -1,66 +1,111 @@
...
+
+			const classes = classnames( {
+				'reader-post-card__gallery-item': true,
+				hide: index !== this.state.itemSelected,
```
Comment (roo2) [inline client/blocks/reader-post-card/gallery.jsx:46]: I imagine we might want to do some kind of css transition in the future? like fading images in and out rather than just show/hide cc @ollierozdarz @davemart-in

File: client/blocks/reader-post-card/gallery.jsx [hunk 9]
```diff
@@ -1,66 +1,111 @@
...
+}
 
 PostGallery.propTypes = {
 	post: PropTypes.object.isRequired,
 	isDiscover: PropTypes.bool,
+	onClick: PropTypes.func,
+};
+
+PostGallery.defaultProps = {
+	onClick: noop,
```
Comment (roo2) [inline client/blocks/reader-post-card/gallery.jsx]: I don't see where `onClick` is used?

File: client/blocks/reader-post-card/style.scss [hunk 10]
```diff
@@ -408,15 +399,15 @@
...
+	flex-direction: column;
 	margin-top: 14px;
 
 	@include breakpoint-deprecated( '<480px' ) {
 		flex-direction: column;
 	}
 
 	.reader-post-card__post-details {
-		flex: 1 auto;
+		//flex: 1 auto;
```
Comment (roo2) [inline client/blocks/reader-post-card/style.scss]: commented out code

File: client/blocks/reader-post-card/style.scss [hunk 11]
```diff
@@ -659,3 +667,26 @@
...
+	padding: 0;
+	margin: 0;
+	flex-direction: row;
+
+	&.column {
+		flex-direction: column;
+	}
+}
+
+/* Create four equal columns that sits next to each other */
```
Comment (roo2) [inline client/blocks/reader-post-card/style.scss]: I think this comment doesn't look related to the CSS below it

**Comment by roo2** [comment]
It looks like the little circles were supposed to be clickable but they aren't, at least in chrome I'm taken to the article:


https://user-images.githubusercontent.com/22446385/189019833-313bd290-6afb-4841-a432-18a3d630a1df.mov

File: client/blocks/reader-post-card/gallery.jsx [hunk 12]
```diff
@@ -1,66 +1,111 @@
...
 		}
+		this.setState( { itemSelected: updateItemSelected } );
+	};
+
+	getGalleryWorthyImages = ( post ) => {
+		const numberOfImagesToDisplay = 10;
+		const images = ( post.images && [ ...post.images ] ) || [];
+
+		return images
+			.sort( ( a, b ) => ( a.width > b.width ? 1 : -1 ) )
```
Comment (roo2) [inline client/blocks/reader-post-card/gallery.jsx]: It looks like this mechanism will show the _largest_ image first, but maybe we should show _the image that appears first in the article_ first (and that is large enough to qualify). What do you think @davemart-in @ollierozdarz ?

**Comment by roo2** [comment]
hmmm this image looks a bit strange. Maybe we should filter out images that are extremely wide or tall as they're likely not suitable to be resampled for the gallery? cc @davemart-in @ollierozdarz 

I'm sure there are many tall or wide images that would work just fine, but I have a feeling that they're generally less likely to be suitable.

source image:
Screen Shot 2022-09-08 at 12 32 50 pm

in gallery:
Screen Shot 2022-09-08 at 12 44 47 pm


EDIT: here's another example of wide images not being suitable for the gallery
Source:
Screen Shot 2022-09-08 at 12 48 30 pm

Galelry:
Screen Shot 2022-09-08 at 12 48 23 pm

**Comment by roo2** [comment]
There seem to be some posts getting duplicate images for some reason. e.g. this feed seems to have the issue on several posts: `/read/feeds/77222666`

Screen Shot 2022-09-08 at 12 52 08 pm

File: client/blocks/reader-post-card/gallery.jsx [hunk 13]
```diff
@@ -1,66 +1,111 @@
...
+					/>
+				</li>
+			);
+		} );
+		const circles = this.state.listItems.map( ( value, index ) => {
+			const classes = classnames( {
+				'reader-post-card__gallery-circle': true,
+				'is-selected': index === this.state.itemSelected,
+			} );
+			return <div className={ classes }></div>;
```
Comment (roo2) [inline client/blocks/reader-post-card/gallery.jsx]: I'm getting a warning in the console, I think this is the div that needs a `key` property

**Comment by roo2** [comment]
it looks like I'm getting a JavaScript error in `/read/conversations`:

Screen Shot 2022-09-08 at 1 04 01 pm

**Comment by roo2** [comment]
on a positive note, I think this looks amazingly good when the site has nice images selected 😊


Screen Shot 2022-09-08 at 12 55 33 pm

**Review by roo2 — COMMENTED** [review]
Looking good, I dived in and did some nitpicking on the code!

File: client/blocks/reader-featured-image/index.jsx (hunk 3 above)
Comment (eoigal) [reply client/blocks/reader-featured-image/index.jsx]: It isn't necessary. I'll remove it.

File: client/blocks/reader-featured-image/index.jsx (hunk 2 above)
Comment (eoigal) [reply client/blocks/reader-featured-image/index.jsx]: Good idea, I'll add this.

File: client/blocks/reader-post-card/gallery.jsx (hunk 5 above)
Comment (eoigal) [reply client/blocks/reader-post-card/gallery.jsx]: Yup!

File: client/blocks/reader-post-card/gallery.jsx (hunk 12 above)
Comment (eoigal) [reply client/blocks/reader-post-card/gallery.jsx]: The sort here was to actually filter out smaller images from the list of images in the post. I suspect that I can rewrite this to avoid sorting (it already sorts by how they are placed in the post afaik) and just filter out smaller/less suitable images that won't suit the gallery view.

File: client/blocks/reader-post-card/gallery.jsx (hunk 6 above)
Comment (eoigal) [reply client/blocks/reader-post-card/gallery.jsx:42]: Agree but prefer composition over early refactoring at this point.

File: client/blocks/reader-post-card/featured-images.jsx (hunk 1 above)
Comment (eoigal) [reply client/blocks/reader-post-card/featured-images.jsx]: Removed this as not used

File: client/blocks/reader-featured-image/index.jsx (hunk 2 above)
Comment (eoigal) [reply client/blocks/reader-featured-image/index.jsx]: Good idea, added null safe check

File: client/blocks/reader-featured-image/index.jsx (hunk 3 above)
Comment (eoigal) [reply client/blocks/reader-featured-image/index.jsx]: Not needed, kinda legacy of me trying to get flex containers to work as I'd like

File: client/blocks/reader-post-card/gallery.jsx (hunk 4 above)
Comment (eoigal) [reply client/blocks/reader-post-card/gallery.jsx]: Yeah, good call. I typically leave refactoring until the end but in this case I think ok to move to helper function

File: client/blocks/reader-post-card/gallery.jsx (hunk 5 above)
Comment (eoigal) [reply client/blocks/reader-post-card/gallery.jsx]: 👍

File: client/blocks/reader-post-card/index.jsx (hunk 7 above)
Comment (eoigal) [reply client/blocks/reader-post-card/index.jsx]: I need to figure out how to handle posts classified as photo posts. This was done just to allow me to see one image featured images.

File: client/blocks/reader-post-card/gallery.jsx (hunk 9 above)
Comment (eoigal) [reply client/blocks/reader-post-card/gallery.jsx]: Not sure why I added that!.. removed and fixed issue with clicking circles.

File: client/blocks/reader-post-card/style.scss (hunk 11 above)
Comment (eoigal) [reply client/blocks/reader-post-card/style.scss]: Removed.

File: client/blocks/reader-post-card/gallery.jsx (hunk 13 above)
Comment (eoigal) [reply client/blocks/reader-post-card/gallery.jsx]: Fixed

**Comment by eoigal** [comment]
Thanks @roo2 again for review.

I've addressed a few things based on your feedback;

* I've consolidated how we decide if an image is suitable to display as a featured image in either a flex container or gallery
* That function now has 3 filters. One to remove duplicates, one to remove small images and another to remove any images outside an ideal image aspect ratio (still to decide what that ideal is...)
* I also removed the sort from this function so the images should be ordered by where they are placed in post.
* I've removed some cruft from the code that leaked in while building the components

File: client/blocks/reader-post-card/gallery.jsx (hunk 12 above)
Comment (eoigal) [reply client/blocks/reader-post-card/gallery.jsx]: I've updated this to remove the sort, add a filter to remove duplicates, add a filter to also remove images that are outside an ideal image aspect ratio.

**Comment by davemart-in** [comment]
@eoigal I did another quick review. When you're ready, I'd say let's just merge this back into #66650 and then we can do any additional bug fixes or fine tuning directly in that PR.

**Comment by davemart-in** [comment]
@eoigal I went ahead and [updated the transition animation](https://github.com/Automattic/wp-calypso/pull/67387/commits/8fe67c58af43f02687851ed9798aff424f091ae1) when going from one gallery image to another.

### Before

CleanShot 2022-09-09 at 11 42 16

### After

CleanShot 2022-09-09 at 11 44 05

File: client/blocks/reader-post-card/gallery.jsx (hunk 12 above)
Comment (roo2) [reply client/blocks/reader-post-card/gallery.jsx]: EDIT: oops I was looking at an old version 👍

File: client/blocks/reader-post-card/gallery.jsx (hunk 6 above)
Comment (roo2) [reply client/blocks/reader-post-card/gallery.jsx:42]: that makes sense

File: client/blocks/reader-post-card/style.scss [hunk 14]
```diff
@@ -595,13 +595,19 @@
...
 	margin: 0 0 17px;
 	padding: 0;
+	position: relative;
 }
 
 .reader-post-card__gallery-item {
 	border: 1px solid var( --color-neutral-0 );
 	cursor: pointer;
 	flex: 1;
+	height: 300px;
```
Comment (roo2) [inline client/blocks/reader-post-card/style.scss]: I haven't tested this but will the `reader-post-card__gallery-item` always be 598x300, even on mobile?

**Comment by roo2** [comment]
nice work @davemart-in it's those little touches 👨‍🍳

**Comment by roo2** [comment]
It looks like the carousel feature has an issue on mobile:

https://user-images.githubusercontent.com/22446385/189651733-69bcf3da-d1c3-4ef4-b7c2-3a3871241229.mov

It looked suspicious that we are now setting the height and width on those images, we probably just need a different CSS setting for mobile.

cc @davemart-in

**Comment by roo2** [comment]
Also an image on the `/read/conversations` tab looks mis-sized
Screen Shot 2022-09-12 at 10 21 19 pm

**Comment by davemart-in** [comment]
I'll take a look at both of those this morning. Thanks for highlighting @roo2!

**Comment by roo2** [comment]
No worries! one more thing actually, it looks like the images are jumping slightly on transition? I noticed this on the `/read/search` page

https://user-images.githubusercontent.com/22446385/189653140-afddb090-b9e2-4f45-87a0-6afd92ba4262.mov

File: client/blocks/reader-post-card/style.scss (hunk 14 above)
Comment (davemart-in) [reply client/blocks/reader-post-card/style.scss]: Updated this in https://github.com/Automattic/wp-calypso/pull/67387/commits/2b98ad641eced13ce7b771c1b937ab92d48184f6

**Comment by davemart-in** [comment]
**Updates**

1) In [this commit](https://github.com/Automattic/wp-calypso/pull/67387/commits/2b98ad641eced13ce7b771c1b937ab92d48184f6) I fixed the gallery issue where images disappear on mobile, addressing [Eru's issue reported here](https://github.com/Automattic/wp-calypso/pull/67387#issuecomment-1243658345). Here's a preview of the fixed state:

CleanShot 2022-09-13 at 13 41 30

2) In [this commit](https://github.com/Automattic/wp-calypso/pull/67387/commits/f304574886463276e76414e10df0b058408ba328) I fixed the issue with malformed images in conversations, which addresses [this issue posted by Eru](https://github.com/Automattic/wp-calypso/pull/67387#issuecomment-1243659873). Here's a preview:

CleanShot 2022-09-13 at 13 53 18@2x

3) @roo2 in [this issue](https://github.com/Automattic/wp-calypso/pull/67387#issuecomment-1243666916) you mentioned a slight jump. I was unable to replicate this. I may have fixed it with the changes above. Mind double checking?

cc @eoigal
